### ISSUE TITLE
Vagrant: mount.nfs: requested NFS version or transport protocol is not supported

### ISSUE THREAD

**Issue by naught101** [mastodon/mastodon#22388]
### Steps to reproduce the problem

Running the steps in https://docs.joinmastodon.org/dev/setup/

`vagant up` fails with:

```
==> default: Mounting NFS shared folders...
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -o vers=3,udp,rw,actimeo=1 192.168.42.1:/home/nedcr/Downloads/source_code/mastodon /vagrant

Stdout from the command:



Stderr from the command:

mount.nfs: requested NFS version or transport protocol is not supported
```

### Expected behaviour

Vagrant up completes

### Actual behaviour

Vagrant up does not complete

### Specifications

Running on Ubuntu 22.04.1 LTS, latest commits on `main` branch.

**Comment by naught101** [comment]
I *think* I fixed it with this (but I'm having other problems, so not 100% sure yet)

```diff
diff --git a/Vagrantfile b/Vagrantfile
index 880cc1849..50233d5b8 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -132,7 +132,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   end
 
   if config.vm.networks.any? { |type, options| type == :private_network }
-    config.vm.synced_folder ".", "/vagrant", type: "nfs", mount_options: ['rw', 'actimeo=1']
+    config.vm.synced_folder ".", "/vagrant", type: "nfs", mount_options: ['rw', 'actimeo=1', 'mountvers=3']
   else
     config.vm.synced_folder ".", "/vagrant"
   end
```

**Comment by naught101** [comment]
But the `.` directory does not sync with `/vagrant`, so later errors occur

**Comment by naught101** [comment]
OK, this change got the syncing working for me:

```diff
diff --git a/Vagrantfile b/Vagrantfile
index 880cc1849..4319638dc 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -131,11 +131,11 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     config.hostsupdater.remove_on_suspend = false
   end
 
-  if config.vm.networks.any? { |type, options| type == :private_network }
-    config.vm.synced_folder ".", "/vagrant", type: "nfs", mount_options: ['rw', 'actimeo=1']
-  else
-    config.vm.synced_folder ".", "/vagrant"
-  end
+  # if config.vm.networks.any? { |type, options| type == :private_network }
+  #   config.vm.synced_folder ".", "/vagrant", type: "nfs", mount_options: ['rw', 'actimeo=1', 'mountvers=4']
+  # else
+  config.vm.synced_folder ".", "/vagrant"
+  # end
 
   # Otherwise, you can access the site at http://localhost:3000 and http://localhost:4000 , http://localhost:8080
   config.vm.network :forwarded_port, guest: 3000, host: 3000
```

**Comment by naught101** [comment]
After running the above and then `vagrant ssh -c "cd /vagrant && bundle install"` (puma and sidekiq were missing), I now have a working dev environment, I think.

Mastodon dev still wasn't working - I needed to run `vagrant ssh -c "cd /vagrant && bin/rails db:migrate RAILS_ENV=development"`

And then the `foreman start` works, and I have a running instance.

**Comment by naught101** [comment]
But the credentials supplied in the dev setup docs don't work (admin@mastodon.local : mastodonadmin)...

**Comment by techhead** [comment]
This issue is related to PR #22075, which sought to fix issue #22074.

**Comment by naught101** [comment]
@techhead but that is already merged 2 weeks ago? I am running mastodon main from yesterday.

**Comment by naught101** [comment]
I needed to run `vagrant reload` after the vagrant file change, too.

**Comment by techhead** [comment]
> @techhead but that is already merged 2 weeks ago? I am running mastodon main from yesterday.

Yes, I understand that. What I'm saying is that part of my fix appears to also be part of your problem. It seems that we are having difficulty in finding NFS mount options that work well across platforms.

See here: https://github.com/mastodon/mastodon/pull/22075#pullrequestreview-1205077271

**Comment by felicianotech** [comment]
I had this issue as well. For me, from what was default, I changed it to `'vers=4,tcp'` to get it to work. I was running this on an Ubuntu 22.04 host.

Three possible solutions I see here:

1. Keep playing around with settings, hopefully get more people to post what works here, and we can find a common denominator.
2. Separate Vagrantfiles for different OSs.
3. Stop using NFS. Personally, this is the first time I had a Vagrantfile need to use one.

**Comment by k3yss** [comment]
Hey, @felicianotech @naught101 were you able to run the development environment? After commenting out the NFS in VagrantFile, I was faced with some dependencies issues which I fixed after ssh into the vagrant instance but at last, I am facing `LetterOpenerWeb` error which looked like a function was not properly called, can you tell me what were the exacts steps that I should follow to get a mastodon instance up and running using Vagrant?

**Comment by kvuzet** [comment]
> But the credentials supplied in the dev setup docs don't work ([admin@mastodon.local](mailto:admin@mastodon.local) : mastodonadmin)...

For anyone looking at this issue trying to get a dev env up on debian, I also had to run `vagrant ssh -c "cd /vagrant && bin/rails db:seed RAILS_ENV=development"` to get the local accounts set up

To summarize this is what I had to do to get it working on debian:
- comment out the lines mentioned in https://github.com/mastodon/mastodon/issues/22388#issuecomment-1356059185
- `vagrant up` (or if already up, run `vagrant reload`)
- `vagrant ssh -c "cd /vagrant && bundle install"`
- `vagrant ssh -c "cd /vagrant && yarn install"`
- `vagrant ssh -c "cd /vagrant && bin/rails db:migrate RAILS_ENV=development"`
- `vagrant ssh -c "cd /vagrant && bin/rails db:seed RAILS_ENV=development"`
- `vagrant ssh -c "cd /vagrant && foreman start"`

**Comment by nickjuntilla** [comment]
For the last line I needed to run `vagrant ssh -c "foreman start -d /vagrant"` instead

**Comment by nickjuntilla** [comment]
None of these settings are working for me to get access to the drive. I get:

```
==> default: Exporting NFS shared folders...
NFS is reporting that your exports file is invalid. Vagrant does
this check before making any changes to the file. Please correct
the issues below and execute "vagrant reload":

exports:2: sandbox_check failed. nfsd has no read access to "/Users/username/Documents/ownerfy/social"
```

**Comment by YourAutisticLife** [comment]
I'll note that I've built v4.2.0-beta1 for vagrant and the only thing I need now is @felicianotech's solution to change the mount line. Everything else is unnecessary.

**Comment by mjankowski** [comment]
Closing on assumption that the previous comment here indicates this is resolved ... but if that's not the case, can you re-confirm issues on latest main branch code and report back here?
